### PR TITLE
Add and update certification type for a tax return

### PR DESCRIPTION
**Pull request by bytheway875**
THIS WAS BRANCHED FROM THE INTERVIEW TIMING PREFERENCE CODE, SO PLEASE REVIEW THAT PR FIRST!!

No Javascript -- uses query params to detect whether an edit has been initiated.

Reviewed design implementation with Anu + Yvonne.

Screen Shot 2020-11-23 at 11 56 02 AM
Screen Shot 2020-11-23 at 11 58 32 AM
Screen Shot 2020-11-23 at 11 58 39 AM
Screen Shot 2020-11-23 at 12 09 44 PM

### CODE REVIEW

File: app/assets/stylesheets/components/_index-table.scss [hunk 1]
```diff
@@ -28,6 +28,10 @@ It expects to be the primary content of the given page (full width).
 
   // for nested tables
   .index-table__cell {
+    padding-right: 1.6rem;
+    &:last-of-type {
+      padding-right: 0;
+    }
```
Comment (bytheway875) [inline app/assets/stylesheets/components/_index-table.scss:34]: Decreases amount of padding between cells on index table, and makes the last cell sit flush with the end of the table.

File: app/assets/stylesheets/components/_tax-return-list.scss [hunk 2]
```diff
@@ -1,5 +1,28 @@
 .tax-return-list {
+  font-size: 1.4rem;
   margin: 0;
+
+  &__certification {
```
Comment (alaroia-cfa) [inline app/assets/stylesheets/components/_tax-return-list.scss:5]: Perhaps you can remove the selector here. Not a blocker.

File: app/controllers/hub/tax_returns/certifications_controller.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  module TaxReturns
+    class CertificationsController < ApplicationController
+      include AccessControllable
+      before_action :require_sign_in
+      load_and_authorize_resource :tax_return, parent: false
+
+      def update
+        @tax_return.update(tax_return_params)
+        params[:tax_return_id] = nil
+        redirect_to referer_without_params || hub_client_path(id: @tax_return.client.id)
```
Comment (alaroia-cfa) [inline app/controllers/hub/tax_returns/certifications_controller.rb:11]: Personally, I'd prefer to see an explicit `?next=...` URL passed in as a query string param.

Make sure it contains no `//` which could be used to switch host, and bob's your uncle, as they say in Australia.

Not a blocker; the referrer-based way is OK too.

File: app/helpers/tax_return_status_helper.rb [hunk 4]
```diff
@@ -19,18 +19,47 @@ def status_translation(status)
...
 
+  def certification_label(tax_return)
+    text = ""
+    classes = ["label", "certification-label"]
+    if tax_return.certification_level.blank?
+      text << "N/A"
+      classes << "unassigned"
+    end
+    if tax_return.certification_level == "advanced"
+      text << "ADV"
```
Comment (alaroia-cfa) [inline app/helpers/tax_return_status_helper.rb:49]: Maybe we should localize these. Not a blocker really.

File: app/views/hub/clients/index.html.erb [hunk 5]
```diff
@@ -26,7 +26,7 @@
         <th scope="col" class="index-table__header">
           <%= render "shared/column_sort_link", title: t("general.updated_at"), column_name: "updated_at" %>
         </th>
-        <th scope="col" class="index-table__header"><%= t(".filing_year") %>, <%= t("general.assignee") %>, <%= t("general.status") %></th>
+        <th scope="col" class="index-table__header" style="width: 35%"><%= t(".filing_year") %>, <%= t("general.assignee") %>, <%= t("general.status") %></th>
```
Comment (alaroia-cfa) [inline app/views/hub/clients/index.html.erb:29]: sgtm

File: app/views/shared/_tax_return_certification_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,13 @@
+<%= form_with class: tax_return, url: update_certification_hub_tax_return_path(id: tax_return.id), local: true, method: :patch, builder: VitaMinFormBuilder, html: { class: "tax-return-certification-form" } do |f| %>
+  <div class="before">
+    <label for="certification_level">Certification</label>
+    <%= link_to "Cancel", url_for(:locale => params[:locale]) %>
+  </div>
+  <%= f.select :certification_level, certification_options_for_select, include_blank: true, selected: tax_return.certification_level %>
+  <%= f.check_box :is_hsa, checked: tax_return.is_hsa? %>HSA?
+  <div>
+    <%= f.button :submit do %>
+      <%= image_tag("checkmark.svg") %>
```
Comment (alaroia-cfa) [inline app/views/shared/_tax_return_certification_form.html.erb:10]: Make accessible, presumably with `alt=""`... unless it doesn't say "Submit" to screen readers

File: app/views/shared/_tax_return_certification_form.html.erb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<%= form_with class: tax_return, url: update_certification_hub_tax_return_path(id: tax_return.id), local: true, method: :patch, builder: VitaMinFormBuilder, html: { class: "tax-return-certification-form" } do |f| %>
+  <div class="before">
+    <label for="certification_level">Certification</label>
+    <%= link_to "Cancel", url_for(:locale => params[:locale]) %>
+  </div>
+  <%= f.select :certification_level, certification_options_for_select, include_blank: true, selected: tax_return.certification_level %>
+  <%= f.check_box :is_hsa, checked: tax_return.is_hsa? %>HSA?
```
Comment (alaroia-cfa) [inline app/views/shared/_tax_return_certification_form.html.erb:7]: i18n HSA

File: app/views/shared/_tax_return_certification_form.html.erb [hunk 8]
```diff
@@ -0,0 +1,13 @@
+<%= form_with class: tax_return, url: update_certification_hub_tax_return_path(id: tax_return.id), local: true, method: :patch, builder: VitaMinFormBuilder, html: { class: "tax-return-certification-form" } do |f| %>
+  <div class="before">
+    <label for="certification_level">Certification</label>
```
Comment (alaroia-cfa) [inline app/views/shared/_tax_return_certification_form.html.erb:3]: I18N Certification

File: app/views/shared/_tax_return_list.html.erb [hunk 9]
```diff
@@ -14,13 +14,30 @@
           <% end %>
         </div>
       </div>
+      <div class="tax-return-list__certification">
+        <% if params[:update_cert] && params[:tax_return_id]&.to_i == tax_return.id %>
+          <%= render "shared/tax_return_certification_form", tax_return: tax_return %>
+        <% else %>
+          <% if status_updateable %>
+            <%= link_to update_cert:true, tax_return_id:tax_return.id do %>
```
Comment (alaroia-cfa) [inline app/views/shared/_tax_return_list.html.erb:22]: `:tax_return.id` -> `: tax_return.id` (add space)

File: db/schema.rb [hunk 10]
```diff
@@ -398,6 +398,16 @@
     t.index ["vita_partner_id"], name: "index_intakes_on_vita_partner_id"
   end
 
+  create_table "memberships", force: :cascade do |t|
```
Comment (alaroia-cfa) [inline db/schema.rb:401]: Maybe not necessary for this PR

File: db/schema.rb [hunk 11]
```diff
@@ -613,6 +625,8 @@
   add_foreign_key "incoming_text_messages", "clients"
   add_foreign_key "intake_site_drop_offs", "intake_site_drop_offs", column: "prior_drop_off_id"
   add_foreign_key "intakes", "vita_partners"
+  add_foreign_key "memberships", "users"
```
Comment (alaroia-cfa) [inline db/schema.rb:628]: Probably not needed for this PR

File: spec/controllers/hub/tax_returns/certifications_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        expect {
+          patch :update, params: params
+          tax_return.reload
+        }.to change(tax_return, :is_hsa).to(true)
+      end
+
+      context "redirecting on success" do
+        context "with referer header" do
+          before do
+            request.headers["Referer"] = "/en/some/path?some=param"
```
Comment (alaroia-cfa) [inline spec/controllers/hub/tax_returns/certifications_controller_spec.rb:34]: FWIW a typical Referer header will be a full URL with hostname.

File: app/helpers/tax_return_status_helper.rb [hunk 13]
```diff
@@ -19,18 +19,47 @@ def status_translation(status)
...
+    end
+    if tax_return.certification_level == "basic"
+      text << "BAS"
+    end
+    text << " | HSA" if tax_return.certification_level.present? && tax_return.is_hsa?
```
Comment (alaroia-cfa) [inline app/helpers/tax_return_status_helper.rb:54]: Let's make "N/A | HSA" possible

**Review by alaroia-cfa — APPROVED** [review]
This is good! Some non-blocking suggestions. It does need the merge conflict resolved :D

**Pull request merged**